azkaban-common/src/main/java/azkaban/imagemgmt/rampup/ImageRampupManagerImpl.java

```diff
@@ -0,0 +1,190 @@
+        .fetchRampupByImageTypes(imageTypes);
+    return this.processAndGetVersionForImageTypes(imageTypes, imageTypeRampups);
+  }
+
+  /**
+   * This method processes image type rampup details for the image type and selects a version for
+   * the image type. Here is the version selection process for an image type 1. Sort the ramp up
+   * data in the ascending order of rampup percentage. 2. Generate a random number between 1 to 100
+   * both inclusive. Let us say the number the number generated is 60. 3. Let us say there are three
+   * versions 1.1.1, 1.1.2 & and 1.1.3 with rampup percantages 10, 30 & 60 respectively. 4. The
+   * above percentage fors three ranges [1 - 10], [11 - 40] & [41 - 100]. The random humber 60
+   * belongs to the last range i.e. [41 - 100] and hence version 1.1.3 will be selected. If random
+   * number generated is 22 then version 1.1.2 will be selected and so on. 5. If there is no active
+   * rampup plan for an image type or in the active plan if the version is marked unstable or
+   * deprecated, and latest active version will be selected from the image_verions table. 6. If
+   * there is no active version in the image_versions table, it will throw appropriate error message
+   * mentioning could not select version for the image type and the whole process would fail. 7.
+   * Follow the rampup procedure to elect a new version from the image_versions table for the ailed
+   * image type.
+   *
+   * @param imageTypes       - set of specified image types
+   * @param imageTypeRampups - contains rampup list for an image type
+   * @return Map<String, String>
+   */
+  private Map<String, String> processAndGetVersionForImageTypes(Set<String> imageTypes,
+      Map<String, List<ImageRampup>> imageTypeRampups) {
+    Set<String> imageTypeSet = imageTypeRampups.keySet();
+    log.info("Found active rampup for the image types {} ", imageTypeSet);
+    Iterator<String> iterator = imageTypeSet.iterator();
+    Map<String, String> imageTypeVersionMap = new TreeMap<>();
+    while (iterator.hasNext()) {
+      String imageTypeName = iterator.next();
+      List<ImageRampup> imageRampupList = imageTypeRampups.get(imageTypeName);
+      Collections.sort(imageRampupList, this.getRampupPercentageComparator());
+      int prevRampupPercentage = 0;
+      int nextRandom = this.getRandomNumberInRange(1, 100);
+      for (ImageRampup imageRampup : imageRampupList) {
+        int rampupPercentage = imageRampup.getRampupPercentage();
+        if (nextRandom >= prevRampupPercentage + 1
+            && nextRandom <= prevRampupPercentage + rampupPercentage) {
+          imageTypeVersionMap.put(imageTypeName, imageRampup.getImageVersion());
+          log.info("The image version {} is selected for image type {} with rampup percentage {}",
+              imageRampup.getImageVersion(), imageTypeName, rampupPercentage);
+          break;
+        }
+        prevRampupPercentage = rampupPercentage;
```

Comment:
I think there is a bug here. The last line should be: prevRampupPercentage += rampupPercentage;